dialog/file_xdg_flatpak.go

```diff
@@ -107,33 +106,84 @@ func fileSaveOSOverride(d *FileDialog) bool {
 	if d.startingLocation != nil {
 		options.CurrentFolder = d.startingLocation.Path()
 	}
+	options.Filters, options.CurrentFilter = convertFilterForPortal(d.filter)
 
 	callback := d.callback.(func(fyne.URIWriteCloser, error))
 	windowHandle := windowHandleForPortal(d.parent)
 
 	go func() {
 		file, err := saveFile(windowHandle, options)
-		fyne.DoAndWait(func() {
+		fyne.Do(func() {
 			callback(file, err)
 		})
 	}()
 
 	return true
 }
 
-func x11WindowHandleToString(handle uintptr) string {
-	return "x11:" + strconv.FormatUint(uint64(handle), 16)
-}
-
 func windowHandleForPortal(window fyne.Window) string {
 	windowHandle := ""
 	if !build.IsWayland {
 		window.(driver.NativeWindow).RunNative(func(context any) {
 			handle := context.(driver.X11WindowContext).WindowHandle
-			windowHandle = x11WindowHandleToString(handle)
+			windowHandle = portal.FormatX11WindowHandle(handle)
 		})
 	}
 
 	// TODO: We need to get the Wayland handle from the xdg_foreign protocol and convert to string on the form "wayland:{id}".
 	return windowHandle
 }
+
+func convertFilterForPortal(fyneFilter storage.FileFilter) (list []*filechooser.Filter, current *filechooser.Filter) {
+	if fyneFilter == nil || fyneFilter == folderFilter {
+		return nil, nil
+	}
+
+	if filter, ok := fyneFilter.(*storage.ExtensionFileFilter); ok {
+		rules := make([]filechooser.Rule, 0, 2*len(filter.Extensions))
+		for _, ext := range filter.Extensions {
+			lowercase := filechooser.Rule{
+				Type:    filechooser.GlobPattern,
+				Pattern: "*" + strings.ToLower(ext),
+			}
+			uppercase := filechooser.Rule{
+				Type:    filechooser.GlobPattern,
+				Pattern: "*" + strings.ToUpper(ext),
+			}
+			rules = append(rules, lowercase, uppercase)
+		}
+
+		name := formatFilterName(filter.Extensions, 3)
+		converted := &filechooser.Filter{Name: name, Rules: rules}
+		return []*filechooser.Filter{converted}, converted
+	}
+
+	if filter, ok := fyneFilter.(*storage.MimeTypeFileFilter); ok {
+		rules := make([]filechooser.Rule, len(filter.MimeTypes))
+		for i, mime := range filter.MimeTypes {
+			rules[i] = filechooser.Rule{
+				Type:    filechooser.MIMEType,
+				Pattern: mime,
+			}
+		}
+
+		name := formatFilterName(filter.MimeTypes, 3)
+		converted := &filechooser.Filter{Name: name, Rules: rules}
+		return []*filechooser.Filter{converted}, converted
+	}
+
+	return nil, nil
+}
+
+func formatFilterName(patterns []string, count int) string {
```

Comment:
I think this would benefit from a unit test.

I'm pretty sure that if you pass ["1", "2", "3"] and count=4 the output won't be correct.

Reply:
Yup. Good catch. Fixed now